### PR TITLE
Alfoa/simulated annealing discrete variables

### PR DESCRIPTION
**Pull request by alfoa**
--------
Pull Request Description
--------
##### What are the significant changes in functionality due to this change request?

----------------
For Change Control Board: Change Request Review
----------------
The following review must be completed by an authorized member of the Change Control Board.
- [ ] 1. Review all computer code.
- [ ] 2. If any changes occur to the input syntax, there must be an accompanying change to the user manual and xsd schema. If the input syntax change deprecates existing input files, a conversion script needs to be added (see Conversion Scripts).
- [ ] 3. Make sure the Python code and commenting standards are respected (camelBack, etc.) - See on the [wiki](https://github.com/idaholab/raven/wiki/RAVEN-Code-Standards#python) for details.
- [ ] 4. Automated Tests should pass, including run_tests, pylint, manual building and xsd tests. If there are changes to Simulation.py or JobHandler.py the qsub tests must pass.
- [ ] 5. If significant functionality is added, there must  be tests added to check this. Tests should cover all possible options.  Multiple short tests are preferred over one large test. If new development on the internal JobHandler parallel system is performed, a cluster test must be added setting, in <RunInfo> XML block, the node ```<internalParallel>``` to True.
- [ ] 6. If the change modifies or adds a requirement or a requirement based test case, the Change Control Board's Chair or designee also needs to approve the change.  The requirements and the requirements test shall be in sync.
- [ ] 7. The merge request must reference an issue.  If the issue is closed, the issue close checklist shall be done.
- [ ] 8. If an analytic test is changed/added is the the analytic documentation updated/added?
- [ ] 9. If any test used as a basis for documentation examples (currently found in `raven/tests/framework/user_guide` and `raven/docs/workshop`) have been changed, the associated documentation must be reviewed and assured the text matches the example.

### CODE REVIEW

File: ravenframework/Optimizers/SimulatedAnnealing.py [hunk 1]
```diff
@@ -661,16 +690,16 @@ def _nextNeighbour(self, rlz, fraction=1):
...
-          delta = \\frac{-amp}{2} + amp * r
+          delta = \\frac{-amp}{2} + amp * r * (upper-lower)
 
       where :math: `r \sim \mathcal{U}(0,1)`
 
       for boltzmann cooling:
       .. math::
 
-          amp = min(\\sqrt(T), \\frac{1}{3*alpha}
+          amp = min(\\sqrt(T), \\frac{upper-lower}{3*learning_rate}
```
Comment (alfoa) [inline ravenframework/Optimizers/SimulatedAnnealing.py:700]:
```suggestion
          amp = min(\\sqrt(T), \\frac{1}{3*learning_rate}
```

File: ravenframework/Optimizers/SimulatedAnnealing.py [hunk 2]
```diff
@@ -679,7 +708,7 @@ def _nextNeighbour(self, rlz, fraction=1):
 
           amp = r
 
-          delta = alpha * T * tan(amp)
+          delta = learning_rate * T * tan(amp) * (upper-lower)
```
Comment (alfoa) [inline ravenframework/Optimizers/SimulatedAnnealing.py:711]:
```suggestion
          delta = learning_rate * T * tan(amp) 
```

File: ravenframework/Optimizers/SimulatedAnnealing.py [hunk 3]
```diff
@@ -688,29 +717,35 @@ def _nextNeighbour(self, rlz, fraction=1):
 
           amp = r
 
-          delta = \\sign(amp-0.5)*T*((1.0+\\frac{1.0}{T})^{\\abs{2*amp-1}-1.0}
+          delta = \\sign(amp-0.5)*T*((1.0+\\frac{1.0}{T})^{\\abs{2*amp-1}-1.0} * (upper-lower)
```
Comment (alfoa) [inline ravenframework/Optimizers/SimulatedAnnealing.py:720]:
```suggestion
          delta = \\sign(amp-0.5)*T*((1.0+\\frac{1.0}{T})^{\\abs{2*amp-1}-1.0}
```

File: ravenframework/Optimizers/SimulatedAnnealing.py [hunk 4]
```diff
@@ -248,9 +250,28 @@ def initialize(self, externalSeeding=None, solutionExport=None):
...
+          val = init[var]
+          values[var] = self.distDict[var].ppf(self.distDict[var].cdf(val))
+          if not  np.isclose(initialValues[traj][var], values[var], 1e-4):
+            self.raiseAWarning(f"Traj: {traj}. Variable {var} is associated with a discrete distribution. The inputted initial value {initialValues[traj][var]} "
+                               f"is not among available discrete values. Closest value is {values[var]}")
+        else:
+          values[var] =  init[var]
+      values = self.normalizeData(values)
+
+      self._submitRun(values,traj,self.getIteration(traj))
```
Comment (wangcj05) [inline ravenframework/Optimizers/SimulatedAnnealing.py:270]: It seems to me these lines can be moved to Optimizer base class.

File: ravenframework/Optimizers/SimulatedAnnealing.py [hunk 5]
```diff
@@ -248,9 +250,28 @@ def initialize(self, externalSeeding=None, solutionExport=None):
...
+
+    # queue up the first run for each trajectory
+    #for traj, init in enumerate(self._initialValues):
+    #  self._submitRun(init,traj,self.getIteration(traj))
```
Comment (wangcj05) [inline ravenframework/Optimizers/SimulatedAnnealing.py:274]: remove?

File: ravenframework/Optimizers/SimulatedAnnealing.py [hunk 6]
```diff
@@ -694,23 +723,29 @@ def _nextNeighbour(self, rlz, fraction=1):
...
     delta = np.atleast_1d(delta)
+
     for i,var in enumerate(self.toBeSampled.keys()):
-      nextNeighbour[var] = rlz[var] + delta[i]
+      nextNeighbour[var] = rlzNormalized[var] + delta[i]
+      nextNeighbour[var] =  0 if nextNeighbour[var] < 0 else nextNeighbour[var]
+      nextNeighbour[var] =  1 if nextNeighbour[var] > 1 else nextNeighbour[var]
+      if self.distDict[var].distType == distType.discrete:
+        val = nextNeighbour[var]
+        nextNeighbour[var] = self.distDict[var].cdf(self.distDict[var].ppf(val))
```
Comment (wangcj05) [inline ravenframework/Optimizers/SimulatedAnnealing.py:748]: Do you need to denormalize the data before you compute ppf?

**Review by wangcj05 — CHANGES_REQUESTED** [review]
@alfoa I have a couple of comments for you to consider. Thanks for the updates regarding the normalized and unnormalized data. It is really confusing in our current code. Some suggestions here: 
1. It seems some of the discrete handling can be moved to Optimizer base class 
2. We need a better document for the data we store in the self. variables. For example, self._optPointHistory stores the normalized data. it is really not clear when our 'rlz' are normalized or not. Again, thanks for you update of our docstring to make it more clear.

File: ravenframework/Optimizers/SimulatedAnnealing.py (hunk 5 above)
Comment (alfoa) [reply ravenframework/Optimizers/SimulatedAnnealing.py:274]: removed

File: ravenframework/Optimizers/SimulatedAnnealing.py (hunk 6 above)
Comment (alfoa) [reply ravenframework/Optimizers/SimulatedAnnealing.py:748]: Theoretically the normalization is performed between 0-1 and it implicitly maps to a uniform distribution. I can add the denormalization in order to avoid that, if in the future the normalization strategy got changed, this would result in a diff

File: ravenframework/Optimizers/SimulatedAnnealing.py (hunk 4 above)
Comment (alfoa) [reply ravenframework/Optimizers/SimulatedAnnealing.py]: I would open an issue to track this in case the RAVEN team is willing to go in that direction

File: ravenframework/Optimizers/SimulatedAnnealing.py (hunk 6 above)
Comment (alfoa) [reply ravenframework/Optimizers/SimulatedAnnealing.py]: added

**Comment by alfoa** [comment]
@wangcj05 all tests fail now for the issue seen in #2317 with  ```cvxpy```

File: ravenframework/Distributions.py [hunk 7]
```diff
@@ -76,6 +77,13 @@ def factorial(x):
                               'UniformDiscrete' : 'UniformDiscreteDistribution'
 }
 
+
+# Declaring namedtuple(DistributionTypes)
+DistributionTypes = namedtuple('DistributionType', ['discrete', 'continuous'])
+# Adding values
+distType = DistributionTypes('Discrete', 'Continuous')
+
+
```
Comment (alfoa) [inline ravenframework/Distributions.py:86]:
```suggestion
```

File: ravenframework/Optimizers/SimulatedAnnealing.py [hunk 8]
```diff
@@ -248,9 +250,28 @@ def initialize(self, externalSeeding=None, solutionExport=None):
...
-      self._submitRun(init,traj,self.getIteration(traj))
+      values = {}
+      init = self.denormalizeData(init)
+      initialValues[traj] = self.denormalizeData(initialValues[traj])
+      for var in init:
+        if var in self.toBeSampled and self.distDict[var].distType == distType.discrete:
+          val = init[var]
+          values[var] = self.distDict[var].ppf(self.distDict[var].cdf(val))
+          if not  np.isclose(initialValues[traj][var], values[var], 1e-4):
+            self.raiseAWarning(f"Traj: {traj}. Variable {var} is associated with a discrete distribution. The inputted initial value {initialValues[traj][var]} "
```
[the remark this answers is not in the document]
Comment (alfoa) [reply ravenframework/Optimizers/SimulatedAnnealing.py:264]: Using the utils.isClose now

File: ravenframework/Distributions.py [hunk 9]
```diff
@@ -1905,6 +1931,7 @@ def initializeDistribution(self):
     paramsDict['state'] = self.pdfArray
 
     self.categoricalDist = Categorical()
+    #isFloat
```
Comment (alfoa) [inline ravenframework/Distributions.py:1934]:
```suggestion
```

File: ravenframework/Optimizers/Optimizer.py [hunk 10]
```diff
@@ -485,10 +485,28 @@ def normalizeData(self, denormed):
 
     return normalized
 
-  def denormalizeData(self, normalized):
+  def denormalizeVariable(self, value, variable):
     """
-      Method to normalize the data
+      Method to enormalize the variable 'variable'
```
Comment (alfoa) [inline ravenframework/Optimizers/Optimizer.py:490]:
```suggestion
      Method to denormalize the variable 'variable'
```

File: ravenframework/Optimizers/Optimizer.py [hunk 11]
```diff
@@ -485,10 +485,28 @@ def normalizeData(self, denormed):
...
+    if variable in self.toBeSampled:
+      lower, upper = self._variableBounds[variable]
+      denormed = value * (upper - lower) + lower
+    return denormed
+
+  def denormalizeData(self, normalized, variable = None):
+    """
+      Method to denormalize the data
       @ In, normalized, dict, dictionary containing the value of decision variables to be denormalized, in form of {varName: varValue}
+      @ In, var, str, optional, if present, only the var 'var' is denormalized
```
Comment (alfoa) [inline ravenframework/Optimizers/Optimizer.py:509]:
```suggestion
```

File: ravenframework/Optimizers/Optimizer.py [hunk 12]
```diff
@@ -485,10 +485,28 @@ def normalizeData(self, denormed):
...
+    # in that case, we allow this method to turn off normalization
+    if self.needDenormalized():
+      return value
+    denormed = value
+    if variable in self.toBeSampled:
+      lower, upper = self._variableBounds[variable]
+      denormed = value * (upper - lower) + lower
+    return denormed
+
+  def denormalizeData(self, normalized, variable = None):
```
Comment (alfoa) [inline ravenframework/Optimizers/Optimizer.py:505]:
```suggestion
  def denormalizeData(self, normalized):
```

File: ravenframework/Optimizers/SimulatedAnnealing.py [hunk 13]
```diff
@@ -248,9 +250,27 @@ def initialize(self, externalSeeding=None, solutionExport=None):
...
+                               f"is not among available discrete values. Closest value is {values[var]}")
+        else:
+          values[var] =  init[var]
+      values = self.normalizeData(values)
+
+      # queue up the first run for each trajectory
+      self._submitRun(values,traj,self.getIteration(traj))
+
+
```
Comment (alfoa) [inline ravenframework/Optimizers/SimulatedAnnealing.py:273]:
```suggestion
```

File: tests/framework/Optimizers/SimulatedAnnealing/continuous/1VariableOptVeryFastSA/test_simulatedAnnealing.xml [hunk 14]
```diff
@@ -59,13 +59,13 @@
...
-        <temperature>1e-2</temperature>
+        <temperature>1e-7</temperature>
         <persistence>1</persistence>
       </convergence>
       <coolingSchedule>
-        <cauchy>
-          <d>1.0</d>
-        </cauchy>
+        <veryfast>
+          <c>0.5</c>
```
Comment (alfoa) [inline tests/framework/Optimizers/SimulatedAnnealing/continuous/1VariableOptVeryFastSA/test_simulatedAnnealing.xml:67]: modifed to reflect the test name "VeryFast" was supposed to be tested here

**Review by wangcj05 — COMMENTED** [review]
minor comment.

File: tests/framework/Optimizers/SimulatedAnnealing/discrete/FunctionallyConstrainedSA/test_funcConstrSimulatedAnnealing.xml [hunk 15]
```diff
@@ -0,0 +1,118 @@
...
+  <TestInfo>
+    <name>framework/Optimizers/SimulatedAnnealing/discrete.FuncionallyConstrainedSA</name>
+    <author>MohammadAbdo</author>
+    <created>2020-02-05</created>
+    <classesTested>Optimizer</classesTested>
+    <description>
+      This test uses a multidimensional linear function such that the trajectory must pass through
+      a functional constraint to reach the optimal point.
+      The test is designed by @talbpaul and is applied to simulated annealing.
+    </description>
```
Comment (alfoa) [inline tests/framework/Optimizers/SimulatedAnnealing/discrete/FunctionallyConstrainedSA/test_funcConstrSimulatedAnnealing.xml:12]:
```suggestion
    <name>framework/Optimizers/SimulatedAnnealing/discrete.FuncionallyConstrainedSA</name>
    <author>alfoa</author>
    <created>2024-05-15</created>
    <classesTested>Optimizer.SimulatedAnnealing</classesTested>
    <description>
     This test assesses the usage of Simulated Annealing with discrete variables and functional constraints. 
     The test is an adaptation of the test ``framework.Optimizers.SimulatedAnnealing.continuous.FuncionallyConstrainedSA'' using discrete distributions.
    </description>
```

[the remark this answers is not in the document]
Comment (alfoa) [reply tests/framework/Optimizers/SimulatedAnnealing/discrete/FunctionallyConstrainedSA/test_funcConstrSimulatedAnnealing.xml]: done

**Review by wangcj05 — APPROVED** [review]
changes are good.
